pip/index.py

```diff
@@ -932,6 +932,15 @@ def egg_fragment(self):
         r'(sha1|sha224|sha384|sha256|sha512|md5)=([a-f0-9]+)'
     )
 
+    _subdirectory_fragment_re = re.compile(r'[#&]subdirectory=([^&]*)')
```

Comment:
Maybe move this after `def hash` method to keep the regex followed by its method pattern.
Maybe also update `_egg_fragment_re = re.compile(r'#egg=([^&]*)')` to `_egg_fragment_re = re.compile(r'[#&]egg=([^&]*)')`